src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -178,103 +173,30 @@ private void GenerateWinMD(MetadataBuilder metadataBuilder, string outputFile)
-                    found |= winrtRules.HasInvalidMethods<ClassDeclarationSyntax>(ref context, publicMethods,classDeclaration.Identifier);
-                }
-
-                foreach (InterfaceDeclarationSyntax interfaceDeclaration in interfaces)
-                {
-                    found |= winrtRules.ImplementsAsyncInterface(ref context, model.GetDeclaredSymbol(interfaceDeclaration), interfaceDeclaration);
-                    
-                    var props = interfaceDeclaration.DescendantNodes().OfType<PropertyDeclarationSyntax>();
-                    found |= winrtRules.CheckPropertySignature(ref context, props, interfaceDeclaration.Identifier);
-                    
-                    var methods = interfaceDeclaration.DescendantNodes().OfType<MethodDeclarationSyntax>();
-                    found |= winrtRules.HasInvalidMethods<InterfaceDeclarationSyntax>(ref context, methods, interfaceDeclaration.Identifier);
-                }
-
-                /* Check all structs */
-                foreach (StructDeclarationSyntax structDeclaration in structs)
-                {
-                    found |= winrtRules.StructHasFieldOfType<ConstructorDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<DelegateDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<EventFieldDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<IndexerDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<MethodDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<OperatorDeclarationSyntax>(ref context, structDeclaration);
-                    found |= winrtRules.StructHasFieldOfType<PropertyDeclarationSyntax>(ref context, structDeclaration);
-
-                    var fields = structDeclaration.DescendantNodes().OfType<FieldDeclarationSyntax>();
-                    foreach (var field in fields) 
-                    { 
-                        found |= winrtRules.CheckFieldValidity(ref context, field, structDeclaration.Identifier, userCreatedTypes); 
-                    }
-                    if (!fields.Any())
-                    {
-                        context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.StructWithNoFieldsRule, structDeclaration.GetLocation(), model.GetDeclaredSymbol(structDeclaration)));
-                        found |= true;
-                    }
-                }
-            }
-            return found;
+            WinRTScanner winrtScanner = new WinRTScanner(context, assemblyName);
+            winrtScanner.FindDiagnostics();
+            return winrtScanner.Found();
         }
 
         public void Execute(GeneratorExecutionContext context)
         {
+            /*

```

Comment:
take this out before final merge, can we use a directive here like `#if NET5_0`

Reply:
tried this and setting it in the diagnostictest project properties didn't seem to apply, so it didn't work...